File: src/core/pipeline/merge_results/handler.py
import json
import os
from collections import defaultdict

import boto3

BATCH_SIZE = 250
_s3 = boto3.client("s3")


def handler(event, context):
    """
    Input:
        event["s3_keys"]:      list of {"s3_key": str, "count": int}
        event["execution_id"]: str

    Returns:
        list of {"s3_key": str, "category": str, "count": int}
    """
    try:
        bucket = os.environ["S3_PIPELINE_BUCKET"]
        s3_refs = event.get("s3_keys", [])
        execution_id = event.get("execution_id", "local")

        all_products: list[dict] = []
        for ref in s3_refs:
            obj = _s3.get_object(Bucket=bucket, Key=ref["s3_key"])
            all_products.extend(json.loads(obj["Body"].read()))
            _s3.delete_object(Bucket=bucket, Key=ref["s3_key"])

        seen_urls: set[str] = set()
        unique: list[dict] = []
        for p in all_products:
            url = p.get("url")
            if url and url not in seen_urls:
                seen_urls.add(url)
                unique.append(p)

        by_category: dict[str, list[dict]] = defaultdict(list)
        for p in unique:
            by_category[p.get("category", "")].append(p)

        batch_refs = []
        index = 0
        for category, products in by_category.items():
            for i in range(0, len(products), BATCH_SIZE):
                batch = products[i : i + BATCH_SIZE]
                s3_key = f"pipeline/batches/{execution_id}/{index}.json"
                _s3.put_object(
                    Bucket=bucket,
                    Key=s3_key,
                    Body=json.dumps(batch),
                    ContentType="application/json",
                )
                batch_refs.append({"s3_key": s3_key, "category": category, "count": len(batch)})
                index += 1

        print(f"[MergeResults] {len(all_products)} total → {len(unique)} unique → {len(batch_refs)} batches ({len(by_category)} categories)")
        return batch_refs

    except Exception as e:
        print(f"[MergeResults] Error: {e}")
        raise

File: src/pipeline/merge-results/src/handler.ts
import { DeleteObjectCommand, GetObjectCommand, PutObjectCommand, S3Client } from '@aws-sdk/client-s3'
import type { IBatchRef, IMergeResultsEvent, IMergeResultsResult, IS3SpiderRef } from './types'

const BATCH_SIZE = 250

let _s3: S3Client | null = null

function getS3(): S3Client {
    if (!_s3) _s3 = new S3Client({})
    return _s3
}

async function loadSpiderResults(s3Refs: IS3SpiderRef[], bucket: string): Promise<Record<string, unknown>[]> {
    const products: Record<string, unknown>[] = []
    for (const ref of s3Refs) {
        const { Body } = await getS3().send(new GetObjectCommand({ Bucket: bucket, Key: ref.s3_key }))
        if (Body) {
            const text = await Body.transformToString()
            products.push(...(JSON.parse(text) as Record<string, unknown>[]))
        }
        await getS3().send(new DeleteObjectCommand({ Bucket: bucket, Key: ref.s3_key }))
    }
    return products
}

function deduplicate(products: Record<string, unknown>[]): Record<string, unknown>[] {
    const seen = new Set<string>()
    return products.filter((p) => {
        const url = p['url'] as string | undefined
        if (!url || seen.has(url)) return false
        seen.add(url)
        return true
    })
}

function groupByCategory(products: Record<string, unknown>[]): Map<string, Record<string, unknown>[]> {
    const groups = new Map<string, Record<string, unknown>[]>()
    for (const p of products) {
        const category = (p['category'] as string | undefined) ?? ''
        if (!groups.has(category)) groups.set(category, [])
        groups.get(category)!.push(p)
    }
    return groups
}

async function writeBatches(groups: Map<string, Record<string, unknown>[]>, executionId: string, bucket: string): Promise<IBatchRef[]> {
    const batchRefs: IBatchRef[] = []
    let index = 0
    for (const [category, products] of groups) {
        for (let i = 0; i < products.length; i += BATCH_SIZE) {
            const batch = products.slice(i, i + BATCH_SIZE)
            const key = `pipeline/batches/${executionId}/${index}.json`
            await getS3().send(
                new PutObjectCommand({
                    Bucket: bucket,
                    Key: key,
                    Body: JSON.stringify(batch),
                    ContentType: 'application/json',
                })
            )
            batchRefs.push({ s3_key: key, category, count: batch.length })
            index++
        }
    }
    return batchRefs
}

export async function handler(event: IMergeResultsEvent): Promise<IMergeResultsResult> {
    const bucket = process.env.S3_PIPELINE_BUCKET
    if (!bucket) throw new Error('Missing required environment variable: S3_PIPELINE_BUCKET')

    const { s3_keys: s3Refs = [], execution_id: executionId = 'local' } = event

    const allProducts = await loadSpiderResults(s3Refs, bucket)
    const unique = deduplicate(allProducts)
    const groups = groupByCategory(unique)
    const batchRefs = await writeBatches(groups, executionId, bucket)

    console.log(`[MergeResults] ${allProducts.length} total → ${unique.length} unique → ${batchRefs.length} batches (${groups.size} categories)`)
    return batchRefs
}
